.github/workflows/test.yaml

```diff
@@ -292,47 +292,82 @@ jobs:
       run: |
           rm -rf "${{ github.workspace }}"
           mkdir "${{ github.workspace }}"
+
     - name: Checkout code
       uses: actions/checkout@v2
       with:
         # spread uses tags as delta reference
         fetch-depth: 0
+
     - name: Cache snapd test results
       id: cache-snapd-test-results
-      uses: actions/cache@v1
+      uses: pat-s/always-upload-cache@v2.1.5
       with:
         path: "${{ github.workspace }}/.test-results"
         key: "${{ github.run_id }}-${{ github.job }}-${{ matrix.system }}-results"
+
     - name: Check cached test results
       id: cached-results
       run: |
-          CACHE_RESULT_STAMP="${{ github.workspace }}/.test-results/${{ matrix.system }}-success"
+          TEST_RESULTS_DIR="${{ github.workspace }}/.test-results"
+          mkdir -p "$TEST_RESULTS_DIR"
+          echo "TEST_RESULTS_DIR=$TEST_RESULTS_DIR" >> $GITHUB_ENV
+
+          CACHE_RESULT_STAMP="$TEST_RESULTS_DIR/${{ matrix.system }}-success"
           echo "CACHE_RESULT_STAMP=$CACHE_RESULT_STAMP" >> $GITHUB_ENV
           if [ -e "$CACHE_RESULT_STAMP" ]; then
               echo "::set-output name=already-ran::true"
           fi
+
+          # Save the var with the failed tests file
+          echo "FAILED_TESTS_FILE=$TEST_RESULTS_DIR/${{ matrix.system }}-failed-tests" >> $GITHUB_ENV
+
+    - name: Check failed tests to run
+      if: "contains(github.event.pull_request.labels.*.name, 'Run failed') && steps.cached-results.outputs.already-ran != 'true'"
+      run: |
+          # Save previous failed test results in FAILED_TESTS env var
+          FAILED_TESTS=""
+          if [ -f "$FAILED_TESTS_FILE" ]; then
+              echo "Failed tests file found"
+              FAILED_TESTS="$(cat $FAILED_TESTS_FILE)"
+          fi
+          if [ -n "$FAILED_TESTS" ]; then
+              echo "Failed tests to run: $FAILED_TESTS"
+              echo "FAILED_TESTS=$FAILED_TESTS" >> $GITHUB_ENV
+          fi
```

Comment:
nitpick: this could be nested in the `if` above

Reply:
yes, updated, thanks